documentation/src/main/asciidoc/user_guide/chapter-63-Streams.adoc

```diff
@@ -0,0 +1,416 @@
+==  Streams
+
+Java 8 introduced the concept of a link:https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html[Stream]
+which allows functional-style operations on collections rather than having to procedurally
+iterate over the data yourself. Stream operations can be implemented in a fashion very
+similar to MapReduce.  Streams, just like MapReduce allow you to perform processing
+upon the entirety of your cache, possibly a very large data set, but in an efficient way.
+
+Also since we can control how the entries are iterated upon we can more efficiently perform the operations in
+a cache that is distributed if you want it to perform all of the operations across the cluster
+concurrently.
+
+A stream is retrieved from the link:https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--[entrySet],
+link:https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--[keySet] or
+link:https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#values--[values] collections returned from the
+Cache by invoking the link:https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#stream--[stream] or
+link:https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#parallelStream--[parallelStream] methods.
+
+=== Common stream operations
+
+This section highlights various options that are present irrespective of what type of underlying cache
+you are using.
+
+==== Key filtering
+
+It is possible to filter the stream so that it only operates upon a given subset of keys.  This can be done
+by invoking the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/CacheStream.html#filterKeys-java.util.Set-[filterKeys]
+method on the `CacheStream`.  This should always be used over a manual filter and will be faster.  Also this
+method allows for distributed computations unlike if you manually retrieve these entries using
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/AdvancedCache.html#getAll-java.util.Set-[getAll]
+and you would have to process locally (pick the best method for the job).
+
+==== Segment based filtering
+
+NOTE: This is an advanced feature and should only be used with deep knowledge of Infinispan segment and hashing techniques.
+
+This option is only supported for replicated and distributed caches.  This allows the user to operate upon
+a subset of data at a time as determined by the
+link:https://docs.jboss.org/infinispan/{infinispanversion}/apidocs/org/infinispan/distribution/DistributionManager.html[DistributionManager].
```

Comment:
Well, the `DistributionManager` (and the consistent hash) just says which segments map to which nodes, the mapping of keys to segments (and so the subset of data) is set in stone in the CH implementation.